Better disk state (#1843)
Nexus-Mods/NexusMods.App: docs/developers/decisions/backend/0016-disk-state-storage.md

# Disk State Storage

## Context

> During the life of a loadout, the application needs to track several the state of the disk. This data includes roughly the

Reviewer: ```suggestion
During the life of a loadout, the application needs to track several states of the disk. This data includes roughly the
```
or "needs to track the state of the disk several times."

following fields:

* Game Path - The path of a file, relative to a LocationId
* Hash - The xxHash64 of the file
* Size - The size of the file
* Last Modified - The last modified date of the file

The latter two entries are not often used directly but are used as shortcuts for detecting if a file needs to be rehashed,
a potentially expensive operation. At any time the application can know that it needs to rehash a file if the size or last
modified date has changed. We include the size in this calculation for use as a failsafe in the case that the user has
modified the file after resetting their system clock.

Additionally, when a loadout is applied to the game, at that point we want to track the state of the disk as the "last applied"
state, that can be used as part of the 3-way compare sync process. In addition, when we manage a game for the first time,
we want to record the state of the disk so that when we create a new loadout we can setup the state of the loadout to match
the unmodified state of the game.

## Important Discoveries During Design

During the design process for this new disk state a few important discoveries were made:

> ### All the disk states for a given game are linear in time

Reviewer: ```suggestion
### All the disk states for a given game installation are linear in time
```

or as `(LoadoutId, LocationId, Path)`, we can perform range queries of `(GameId, ...)` or `(LoadoutId, ...)` to get all the
entities that are relevant to a given game or loadout, sorted by path. This means that we then can perform a 3-way merge
join on the sources of the synchronizer, and remove a lot of the secondary indexing performed in previous implementations. 

In addition, if we query an `IndexSegment` for a given prefix, a binary search can be used to find specific entires in the
segment, reducing the need to scan the entire segment.

Finally, part of the synchronization process is to group loadout files by their location, and then select a winner from any
conflicting items. In the past this required a Linq GroupBy operation that caused a lot of overhead. With all the paths being
loaded pre-sorted, and in a single `IndexSegment`, the group-by can be performed by finding duplicate entries in the segment
and then passing around the group as a `start, end` range in the segment. Essentially this means that grouping can be done
on the stack instead of the heap.

## Implementation

The implementation of the disk state storage is as follows:

> ### GameMetadata

Reviewer: If suggestion to rename this is accepted also change here:
```suggestion
### GameInstallMetadata
```

Each game installation, when detected, causes the creation of a `GameMetadata` entity. We attach disk state to this entity, and
this entity also contains pointers for the following:

> * LastAppliedLoadout - The last loadout that was applied to the game
> * LastAppliedLoadoutTransaction - The transaction that was created when the loadout was applied
> * LastScannedDiskState - The transaction that was created when the disk state was last synced
> * InitialStateTransaction - The transaction that was created when the game was first scanned

Reviewer: I made suggestions to update some of the names here to be more explicit, as the Transaction Type does not otherwise indicate what Entity they should be used as reference for.
```suggestion
* LastAppliedLoadout - The last loadout that was applied to the game
* LastAppliedLoadoutTransaction - The transaction that was created when the loadout was applied
* LastScannedDiskStateTransaction - The transaction that was created when the disk state was last synced
* InitialDiskStateTransaction - The transaction that was created when the game was first scanned
```
